navbar_alerts: Add button to hide "Server Upgrade" alert for 7 days.

Riken-Shah opened the pull request:
Fixes: #18359

This button will help users to avoid a scary red banner
across their screen, while they wait for their sysadmin to
do the upgrade work.

<!-- What's this PR for?  (Just a link to an issue is fine.) -->


**Testing plan:** <!-- How have you tested? -->
Tested in the following states locally:
- Avoid for one week.
- After one week.
- If the local storage value doesn't exist.


**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
![Screenshot 2021-05-05 at 11 08 36 PM](https://user-images.githubusercontent.com/63820270/117188235-7584d180-adfa-11eb-96f3-23b76ab993b2.png)



<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

aryanshridhar commented on static/js/panels.js:
```diff
@@ -83,7 +83,12 @@ export function initialize() {
     } else if (page_params.server_needs_upgrade) {
-        open($("[data-process='server-needs-upgrade']"));
+        const avoid_notification = localstorage.supported() && ls.get("avoidNotificationFor7Days");
+        const seven_days_from_now = Date.now() + 7 * 86400000;
```
Isn't something like: 
```javascript
const seven_days_from_now = getTime(addDays(new Date(), 7));
```
is better than hardcoding the milliseconds ?

Riken-Shah replied:
Thanks, @aryanshridhar!

timabbott commented on templates/zerver/app/navbar_alerts.html:
```diff
@@ -48,9 +48,14 @@
+                    {{ _("Learn more") }}
+                </a>
+                &bull;
+                <a class="alert-link avoid-notification-for-seven-days" role="button" tabindex=0>{{ _("Dismiss for a week") }}</a>
```
A word on naming: We don't want to include the "7 days" duration in variable names and the like, because then it's a pain if we want to change our mind.  Instead we should call this, e.g. `dismiss-upgrade-nag`, and the `seven_days_from_avoid_notification` something like `upgrade_nag_dismissal_duration`.

Riken-Shah replied:
Ahh! Yes. I'll keep this is in mind now.

timabbott commented on static/js/panels.js:
```diff
@@ -83,7 +83,14 @@ export function initialize() {
     } else if (page_params.server_needs_upgrade) {
-        open($("[data-process='server-needs-upgrade']"));
+        const avoid_notification =
+            localstorage.supported() && new Date(ls.get("avoidNotificationFor7Days"));
```
This can be `lastNotificationDismissalTime`

Riken-Shah commented:
@timabbott ready for the review!

timabbott commented on static/js/panels.js:
```diff
@@ -83,7 +84,19 @@ export function initialize() {
+        if (
+            !last_notification_dismissal_time ||
+            !(last_notification_dismissal_time < upgrade_nag_dismissal_duration)
+        ) {
```
I don't understand this logic, and suspect is it is wrong.  We want to be comparing `now` to `last_notification_dismissal_time + 7 days`.  Whereas you seem to be checking something like `x + 7 days > x`.

In addition to fixing that, can you extract this test logic to a function, and write some node tests for it in a new `frontend_tests/node_tests/panels.js`?

Riken-Shah replied:
> I don't understand this logic, and suspect is it is wrong. We want to be comparing now to last_notification_dismissal_time + 7 days. Whereas you seem to be checking something like x + 7 days > x.

Yeah, you are right, What was I thinking! 🤦

Riken-Shah replied:
> In addition to fixing that, can you extract this test logic to a function, and write some node tests for it in a new frontend_tests/node_tests/panels.js?

Should I also add tests for other functions like `should_show_notifications` and `check_profile_incomplete`?

timabbott replied:
Ideally, but if it proves time-consuming, feel free to just do that in a follow-up PR.

timabbott replied:
(This PR is a release-blocker, so I wouldn't add an extra day or two to when it's finished to add tests for code that we've already hand-tested.)

Riken-Shah commented:
@timabbott I fixed the above issue and added node test for `server_upgrade_alert` for now!

I'll try to add the remaining test by tomorrow. But If it takes longer then we can just merge this.

Can you please review the changes (especially node tests)?

timabbott commented:
Merged after several changes:
* Made the testing clearer and more robust -- as you can see, for any time-oriented test, it's a lot cleaner to share the test setup between the true/false code paths so that one can confident that differences in output reflect reality.
* Renamed the local storage parameter for clarity
* Extracted the click handler as a function for convenience of the tests.

Thanks @Riken-Shah!
